circuits/aes-gcm/aes-gcm-foldable.circom

```diff
@@ -24,335 +23,59 @@ include "gctr.circom";
-    signal input aad[16];
-    signal input cipherText[l];
-    signal output blocks[ghashBlocks*4*4];
-
-    // set aad as first block (16 bytes)
-    var blockIndex = 0;
-    for (var i = 0; i<16; i++) {
-        blocks[blockIndex] <== aad[i];
-        blockIndex += 1;
-    }
-
-    // layout blocks of cipherText (l*16 bytes)
-    for (var i=0; i<l; i++) {
-        blocks[blockIndex] <== cipherText[i];
-        blockIndex += 1;
-    }
-
-    // length of aad = 128 = 0x80 as 64 bit number (8 bytes)
-    signal lengthData[8] <== [0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x80];
-    for (var i = 0; i<8; i++) {
-        blocks[blockIndex] <== lengthData[i];
-        blockIndex += 1;
-    }
-
-    // length of blocks as a u64 (8 bytes)
-    var len = totalBlocks * 128;
-    for (var i=0; i<8; i++) {
-        var byteValue = 0;
-        var val = 1;
-        for (var j=0; j<8; j++) {
-            var bit = (len >> i*8+j) & 1;
-            byteValue += bit*val;
-            val = val+val;
-        }
-        // Insert in reversed (big endian) order. 
-        blocks[blockIndex+7-i] <== byteValue;
-    }
-    // 16 + l + 8 + 8
-    blockIndex+=8; // TODO(CR 2024-10-18): I don't think this does anything
-}
-
-// TODO: Mildly more efficient if we add this, maybe it's needed?
-// template GhashStartAndEndMode(l, blockCount, ghashBlocks) {}
-
-template GhashStreamMode(l, ghashBlocks) {
-    signal input cipherText[l];
-    signal output blocks[ghashBlocks*4*4];
-
-    var blockIndex = 0;
-    // layout ciphertext (l*16 bytes)
-    for (var i=0; i<l; i++) {
-        blocks[blockIndex] <== cipherText[i];
-        blockIndex += 1;
-    }
-    
-    // pad remainder 
-    for (var i=blockIndex; i<ghashBlocks*4*4; i++) {
-        blocks[i] <== 0x00; 
-    }
-}
-
-template GhashEndMode(l, totalBlocks, ghashBlocks) {
-    signal input cipherText[l];
-    signal output blocks[ghashBlocks*4*4];
-
-    var blockIndex = 0;
-    // layout ciphertext (l*16 bytes)
-    for (var i=0; i<l; i++) {
-        blocks[blockIndex] <== cipherText[i];
-        blockIndex += 1;
-    }
-
-    signal lengthData[8] <== [0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x80];
-    for (var i = 0; i<8; i++) {
-        blocks[blockIndex] <== lengthData[i];
-        blockIndex += 1;
-    }
-
-    // length of blocks as a u64 (8 bytes)
-    var len = totalBlocks * 128;
-    for (var i=0; i<8; i++) {
-        var byte_value = 0;
-        var val = 1;
-        for (var j=0; j<8; j++) {
-            var bit = (len >> i*8+j) & 1;
-            byte_value += bit*val;
-            val = val+val;
-        }
-        // Insert in reversed (big endian) order. 
-        blocks[blockIndex+7-i] <== byte_value;
-    }
-    blockIndex+=8; 
-    // NOTE: Added this so all of blocks is written
-    for (var i = 0; i<16; i++) {
-        blocks[blockIndex] <== 0;
-        blockIndex += 1;
+    // extract the counter column wise.
+    for (var i = 0; i < 4; i++) {
+        counter[i] <== J0[i][3];
     }
```

Comment:
you love to see a big chunk killed like this. Was this all just stuff we wrote while figuring things out but didn't use?